### PR TITLE
Initial powdr acceleration

### PR DESCRIPTION
**Pull request by leonardoalt**
This PR
- Starts with the https://github.com/powdr-labs/elliptic-curves/tree/powdr branch
- Then uses https://github.com/powdr-labs/elliptic-curves/pull/2
- Adds cfg zkvm guards everywhere so that the native code compiles & its tests run
- Cleans up
- Updates with upstream's master branch

I think we should just merge this PR to our master and keep working on top.

### CODE REVIEW

File: k256/src/arithmetic/hash2curve.rs [hunk 1]
```diff
@@ -415,8 +415,14 @@ mod tests {
             Scalar(reduced_scalar)
         };
 
-        proptest!(ProptestConfig::with_cases(1000), |(b0 in ANY, b1 in ANY, b2 in ANY, b3 in ANY, b4 in ANY, b5 in ANY)| {
-            let mut data = Array::default();
+        proptest!(ProptestConfig::with_cases(
+            if cfg!(all(target_os = "zkvm", target_arch = "riscv32")) {
+                1
```
Comment (leonardoalt) [inline k256/src/arithmetic/hash2curve.rs:420]: @qwang98 why did you change this from 1000 to 1 in your PR?

File: Cargo.lock [hunk 2]
```diff
@@ -796,6 +803,22 @@ version = "1.9.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "cc9c68a3f6da06753e9335d63e27f6b9754dd1920d941135b7ea8224f141adb2"
 
+[[package]]
+name = "powdr-riscv-runtime"
+version = "0.1.0-alpha.2"
```
Comment (pacheco) [inline Cargo.lock:808]: is this right?

File: k256/src/arithmetic/field.rs [hunk 3]
```diff
@@ -5,7 +5,10 @@
 use cfg_if::cfg_if;
 
 cfg_if! {
-    if #[cfg(target_pointer_width = "32")] {
+
+    if #[cfg(all(target_os = "zkvm", target_arch = "riscv32"))] {
+        mod field_8x32_risc0;
```
Comment (pacheco) [inline k256/src/arithmetic/field.rs:10]: powdr?

File: k256/src/arithmetic/field/field_10x26.rs [hunk 4]
```diff
@@ -79,6 +159,17 @@ impl FieldElement10x26 {
         CtOption::new(res, !overflow)
     }
 
+    // /// Attempts to parse the given byte array as an SEC1-encoded field element (but little-endian!).
+    // ///
+    // /// Returns None if the byte array does not contain a big-endian integer in the range
+    // /// [0, p).
+    // pub fn from_bytes_le(bytes: &FieldBytes) -> CtOption<Self> {
```
Comment (pacheco) [inline k256/src/arithmetic/field/field_10x26.rs:166]: leftover maybe?

File: k256/src/arithmetic/hash2curve.rs (hunk 1 above)
Comment (qwang98) [reply k256/src/arithmetic/hash2curve.rs:420]: PR #2 is mix of Risc0 field implementation and Powdr modmul syscall, so this part is actually from the Risc0 commit here: https://github.com/risc0/RustCrypto-elliptic-curves/blob/risczero/k256/src/arithmetic/hash2curve.rs#L418-L424. Other ported over elements are also from this same commit.

File: Cargo.lock (hunk 2 above)
Comment (leonardoalt) [reply Cargo.lock:808]: this is the current version of that package because we didn't update it with the releases. But we actually merged 0.1.2 today, so a cargo update should change this too

File: k256/src/arithmetic/field/field_10x26.rs (hunk 4 above)
Comment (leonardoalt) [reply k256/src/arithmetic/field/field_10x26.rs:166]: yea maybe but it was still there and I didn't write it so didn't want to remove it entirely

File: k256/src/arithmetic/field/field_8x32_risc0.rs [hunk 5]
```diff
@@ -0,0 +1,328 @@
+//! Field element modulo the curve internal modulus using 32-bit limbs.
+#![allow(unsafe_code)]
+
+// This file is the same as the original below except for minor snippets.
+// https://github.com/risc0/RustCrypto-elliptic-curves/blob/risczero/k256/src/arithmetic/field/field_8x32_risc0.rs
```
[the remark this answers is not in the document]
Comment (leonardoalt) [reply k256/src/arithmetic/field/field_8x32_risc0.rs:5]: yea ok. I mean, IMO we could actually only call it `8x32` and keep this attribution, and also not call it powdr

File: k256/src/arithmetic/field.rs (hunk 3 above)
Comment (leonardoalt) [reply k256/src/arithmetic/field.rs:10]: rn I actually prefer just `field_8x32`

File: k256/src/arithmetic/field/field_8x32_risc0.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (leonardoalt) [reply k256/src/arithmetic/field/field_8x32_risc0.rs]: done

File: k256/src/arithmetic/field.rs (hunk 3 above)
Comment (leonardoalt) [reply k256/src/arithmetic/field.rs]: done

File: Cargo.lock (hunk 2 above)
Comment (leonardoalt) [reply Cargo.lock:808]: yea this is actually correct because we use the tag v0.1.1 which had this wrong version

**Comment by leonardoalt** [comment]
review fixed

**Comment by leonardoalt** [comment]
fixed failing native test

**Review by pacheco — APPROVED** [review]
